pkg/datapath/linux/bigtcp/bigtcp.go

```diff
@@ -154,149 +210,157 @@ func probeTSOMaxSize() int {
+	bigv4 := b.UserConfig.EnableIPv4BIGTCP
 
-	modifiedDevices := []string{}
-	for _, device := range option.Config.GetDevices() {
-		// We always add the device because we might do only a partial
-		// modification and end up with an error, so best to be conservative
-		// and always reset all on error.
-		modifiedDevices = append(modifiedDevices, device)
+	log.Infof("Setting up BIG TCP")
+	tsoMax := probeTSOMaxSize(deviceNames)
+	if bigv4 {
+		b.Configuration.groIPv4MaxSize = tsoMax
+		b.Configuration.gsoIPv4MaxSize = tsoMax
+	}
+	if bigv6 {
+		b.Configuration.groIPv6MaxSize = tsoMax
+		b.Configuration.gsoIPv6MaxSize = tsoMax
+	}
+
+	var errs []error
+	for device := range deviceNames {
 		// For compatibility, the kernel will also update the net device's
 		// {gso,gro}_ipv4_max_size, if the new size of {gso,gro}_max_size
 		// isn't greater than 64KB. So it needs to set the IPv6 one first
 		// as otherwise the IPv4 BIG TCP value will be reset.
-		if haveIPv6 {
-			err = setGROGSOIPv6MaxSize(device,
-				bigTCPConfig.groIPv6MaxSize, bigTCPConfig.gsoIPv6MaxSize)
+		if bigv4 {
+			err := setGROGSOIPv6MaxSize(b.UserConfig, device,
+				b.Configuration.groIPv6MaxSize, b.Configuration.gsoIPv6MaxSize)
 			if err != nil {
-				log.WithError(err).WithField("device", device).Warnf("Could not modify IPv6 gro_max_size and gso_max_size%s",
-					disableMsg)
-				option.Config.EnableIPv4BIGTCP = false
-				option.Config.EnableIPv6BIGTCP = false
-				break
+				errs = append(errs, fmt.Errorf("failed to modify IPv6 gro_max_size and gso_max_size for device %q: %w", device, err))
+				continue
 			}
 			log.WithField("device", device).Infof("Setting IPv6 gso_max_size to %d and gro_max_size to %d",
-				bigTCPConfig.gsoIPv6MaxSize, bigTCPConfig.groIPv6MaxSize)
+				b.Configuration.gsoIPv6MaxSize, b.Configuration.groIPv6MaxSize)
 		}
-		if haveIPv4 {
-			err = setGROGSOIPv4MaxSize(device,
-				bigTCPConfig.groIPv4MaxSize, bigTCPConfig.gsoIPv4MaxSize)
+		if bigv6 {
+			err := setGROGSOIPv4MaxSize(b.UserConfig, device,
+				b.Configuration.groIPv4MaxSize, b.Configuration.gsoIPv4MaxSize)
 			if err != nil {
-				log.WithError(err).WithField("device", device).Warnf("Could not modify IPv4 gro_max_size and gso_max_size%s",
-					disableMsg)
-				option.Config.EnableIPv4BIGTCP = false
-				option.Config.EnableIPv6BIGTCP = false
-				break
+				errs = append(errs, fmt.Errorf("failed to modify IPv4 gro_max_size and gso_max_size for device %q: %w", device, err))
+				continue
 			}
 			log.WithField("device", device).Infof("Setting IPv4 gso_max_size to %d and gro_max_size to %d",
-				bigTCPConfig.gsoIPv4MaxSize, bigTCPConfig.groIPv4MaxSize)
-		}
-	}
-
-	if err != nil {
-		if haveIPv4 {
-			bigTCPConfig.groIPv4MaxSize = defaultGROMaxSize
-			bigTCPConfig.gsoIPv4MaxSize = defaultGSOMaxSize
-		}
-		if haveIPv6 {
-			bigTCPConfig.groIPv6MaxSize = defaultGROMaxSize
-			bigTCPConfig.gsoIPv6MaxSize = defaultGSOMaxSize
-		}
-		for _, device := range modifiedDevices {
```

Comment:
@NikAleksandrov @borkmann hey, what I've done in this PR is to switch BIG TCP to reconcile on any changes to the native devices that Cilium is using. In addition to that it's now retrying any failures to set the GRO/GSO, but with a fairly long maximum retry interval. With that, I decided to drop the rollback code completely.

However, I'm having some doubts about whether this is the right behavior. What happens if we persistently fail to set the high GRO/GSO on one of the devices? The retry interval will rise all the way to 30 minutes to avoid log spam, but I wonder if mismatching settings in itself could lead to problems?

Reply:
We'll also follow up with a very similar PR to do these same changes for Bandwidth Manager, so in order to save some time, do you see similar issues there? Anything we should keep in mind? cc @dylandreimerink.

Reply:
> @NikAleksandrov @borkmann hey, what I've done in this PR is to switch BIG TCP to reconcile on any changes to the native devices that Cilium is using. In addition to that it's now retrying any failures to set the GRO/GSO, but with a fairly long maximum retry interval. With that, I decided to drop the rollback code completely.
> 
> However, I'm having some doubts about whether this is the right behavior. What happens if we persistently fail to set the high GRO/GSO on one of the devices? The retry interval will rise all the way to 30 minutes to avoid log spam, but I wonder if mismatching settings in itself could lead to problems?

I haven't tested with HW NIC in terms of what happens if the phys NICs mismatch that of the veth one. Right now it's all aligned, and ideally it should be all aligned. If for some weird reason the limit becomes lower on one of the NICs (I presume this would only happen if a new NIC is dynamically added which would have lower limit), then it would probably make sense to align it also for all NICs in hostns (phys ones, cilium_*, lxc*).

Similarly, if a user turns on the feature where it was off before, the reconciler could detect this and attempt to align the higher limit to all NICs in hostns (phys ones, cilium_*, lxc*).

One thing that is important is the ordering of settings, so first IPv6, then IPv4 as the other way around it would override the IPv4 one.

Reply:
Thanks! Seems like I had missed the veth gro/gso setup in `pkg/datapath/connector/veth.go`. So the reconciler will need to not only deal with native devices, but all lxc and `cilium_*` devices. 

As we're trying to solve the runtime device detection in stages, with the first one just being switching uses of `GetDevices()` to the `Table[*Device]` I'm going to split the reconciliation commits into its own PR and iterate on that there later on.